### PR TITLE
bpo-35721: Close socket pair if Popen in _UnixSubprocessTransport fails

### PR DESCRIPTION
**Pull request by niklasf**
This slightly expands an existing test case `test_popen_error` to trigger a `ResourceWarning` and fixes it.

https://bugs.python.org/issue35721

### CODE REVIEW

File: Lib/asyncio/unix_events.py [hunk 1]
```diff
@@ -757,12 +757,18 @@ def _start(self, args, shell, stdin, stdout, stderr, bufsize, **kwargs):
...
-        if stdin_w is not None:
-            stdin.close()
-            self._proc.stdin = open(stdin_w.detach(), 'wb', buffering=bufsize)
+        try:
+            self._proc = subprocess.Popen(
+                args, shell=shell, stdin=stdin, stdout=stdout, stderr=stderr,
+                universal_newlines=False, bufsize=bufsize, **kwargs)
+            if stdin_w is not None:
+                self._proc.stdin = open(stdin_w.detach(), 'wb', buffering=bufsize)
+                stdin.close()
```
Comment (asvetlov) [inline Lib/asyncio/unix_events.py:765]: Please move the line before `open()` call.
It saves 1 open file descriptor.
Not a very big deal but why not?

File: Lib/test/test_asyncio/test_subprocess.py [hunk 2]
```diff
@@ -480,7 +480,9 @@ def test_popen_error(self):
             popen.side_effect = exc
 
             create = asyncio.create_subprocess_exec(sys.executable, '-c',
-                                                    'pass', loop=self.loop)
+                                                    'pass',
+                                                    stdin=subprocess.PIPE,
```
Comment (asvetlov) [inline Lib/test/test_asyncio/test_subprocess.py]: Please add a new test instead of modifying the existing one.
Two tests for `stdin=PIPE` passed and `None` worth to be tested.
If you don't want to duplicate too much test code extract it into a separate method, like here: https://github.com/python/cpython/blob/master/Lib/test/test_asyncio/test_events.py#L572-L607

**Review by asvetlov — CHANGES_REQUESTED** [review]

File: Lib/asyncio/unix_events.py (hunk 1 above)
Comment (1st1) [reply Lib/asyncio/unix_events.py:765]: I'm not sure I even understand this code :) What's it for?  Why does monkeypatching `self._proc.stdin` work?

Comment (asvetlov) [reply Lib/asyncio/unix_events.py:765]: @1st1 please read a code comment at the begin of the method.
AIX requires to make stdin a socket pair (I love all Unix flavors and their differences).
`proc.stdin` should be a writable file-like object. 
`open()` creates a file object from write-fd of created socket pair.

It exists in asyncio starting from very old times.

Comment (niklasf) [reply Lib/asyncio/unix_events.py:765]: @asvetlov: Good point. I suppose since `open(stdin_w.detach(), ...)` cannot fail I should have kept the order.

Comment (1st1) [reply Lib/asyncio/unix_events.py:765]: > @1st1 please read a code comment at the begin of the method.

Yeah, that's what I did first.  The comment doesn't explain why we pass the *read* end of the pipe to Popen initially (I see that we then monkey-patch it to the *write* end).

> It exists in asyncio starting from very old times.
 
I've no problem with the code or this PR, but I wish that the comment is a bit more elaborate.

**Comment by niklasf** [comment]
I have made the requested changes; please review again.

File: Lib/asyncio/unix_events.py (hunk 1 above)
Comment (asvetlov) [reply Lib/asyncio/unix_events.py:765]: @1st1 you know, I'm bad in comment writing, sorry.
Would you propose exact text?

**Comment by niklasf** [comment]
Is there anything I can do to help this move along?

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
Let's land it. 
Thanks!

**Comment by niklasf** [comment]
Thanks!
